crates/subspace-networking/src/peer_info/handler.rs

```diff
@@ -0,0 +1,281 @@
+        ConnectionHandlerEvent<
+            ReadyUpgrade<&'static [u8]>,
+            Self::OutboundOpenInfo,
+            super::Result,
+            Self::Error,
+        >,
+    > {
+        if let Some(error) = self.error.take() {
+            return Poll::Ready(ConnectionHandlerEvent::Close(error));
+        }
+
+        // Respond to inbound requests.
+        if let Some(fut) = self.inbound.as_mut() {
+            match fut.poll_unpin(cx) {
+                Poll::Pending => {}
+                Poll::Ready(Err(err)) => {
+                    debug!(?err, "Peer info handler: inbound peer info error.");
+
+                    return Poll::Ready(ConnectionHandlerEvent::Close(PeerInfoError::Other {
+                        error: Box::new(err),
+                    }));
+                }
+                Poll::Ready(Ok((stream, peer_info))) => {
+                    debug!(?peer_info, "Inbound peer info");
+
+                    self.inbound = Some(protocol::recv(stream).boxed());
+                    return Poll::Ready(ConnectionHandlerEvent::Custom(Ok(
+                        PeerInfoSuccess::Received(peer_info),
+                    )));
+                }
+            }
+        }
+
+        // Outbound requests.
+        match self.outbound.take() {
+            Some(OutboundState::SendingData(mut peer_info_fut)) => {
+                match peer_info_fut.poll_unpin(cx) {
+                    Poll::Pending => {
+                        self.outbound = Some(OutboundState::SendingData(peer_info_fut));
+                    }
+                    Poll::Ready(Ok(stream)) => {
+                        self.outbound = Some(OutboundState::Idle(stream));
+
+                        return Poll::Ready(ConnectionHandlerEvent::Custom(Ok(
+                            PeerInfoSuccess::DataSent,
+                        )));
+                    }
+                    Poll::Ready(Err(error)) => {
+                        debug!(?error, "Outbound peer info error.",);
+
+                        self.error = Some(PeerInfoError::Other {
+                            error: Box::new(error),
+                        });
+                    }
+                }
+            }
+            Some(OutboundState::Idle(stream)) => {
+                // Nothing to do but we have a negotiated stream.
+                self.outbound = Some(OutboundState::Idle(stream));
+            }
+            Some(OutboundState::NegotiatingStream) => {
+                self.outbound = Some(OutboundState::NegotiatingStream);
+            }
+            Some(OutboundState::RequestNewStream(peer_info)) => {
+                self.outbound = Some(OutboundState::NegotiatingStream);
+                let protocol =
+                    SubstreamProtocol::new(ReadyUpgrade::new(self.config.protocol_name), peer_info)
+                        .with_timeout(self.config.timeout);
+                return Poll::Ready(ConnectionHandlerEvent::OutboundSubstreamRequest { protocol });
+            }
```

Comment:
These for example do not quite look right. The first two do not make any sense, they are no-op and `cx` is not stored, so `poll` will never ever be called again and even if it is by accident. Can you describe why this is the case and how it is supposed to work.